### ISSUE TITLE
Permission denied when running solo-mode

### ISSUE THREAD

**Issue by dawidlenkiewicz** [heroku/heroku-buildpack-nginx#38]
Hello, I'm trying to setup the nginx so it works in solo mode and just redirects the requests to proper app. This is my config:

```
daemon off;
# Heroku dynos have at least 4 cores.
worker_processes <%= ENV['NGINX_WORKERS'] || 4 %>;

events {
  use epoll;
  accept_mutex on;
  worker_connections <%= ENV['NGINX_WORKER_CONNECTIONS'] || 1024 %>;
}

http {
  gzip on;
  gzip_comp_level 2;
  gzip_min_length 512;

  server_tokens off;

  # log_format l2met 'measure#nginx.service=$request_time request_id=$http_x_request_id';
  # access_log <%= ENV['NGINX_ACCESS_LOG_PATH'] || 'logs/nginx/access.log' %> l2met;
  # error_log <%= ENV['NGINX_ERROR_LOG_PATH'] || 'logs/nginx/error.log' %>;

  include mime.types;
  default_type application/octet-stream;
  sendfile on;

  # Must read the body in 5 seconds.
  client_body_timeout <%= ENV['NGINX_CLIENT_BODY_TIMEOUT'] || 5 %>;

  server {
    listen 80;

    location / {
      proxy_pass "main app url";
      proxy_redirect     off;

      proxy_set_header   Host             $host;
      proxy_set_header   X-Real-IP        $remote_addr;
      proxy_set_header   X-Forwarded-For  $proxy_add_x_forwarded_for;

      proxy_connect_timeout      90;
      proxy_send_timeout         90;
      proxy_read_timeout         90;

      proxy_buffer_size          4k;
      proxy_buffers              4 32k;
      proxy_busy_buffers_size    64k;
      proxy_temp_file_write_size 64k;
    }

    location /test {
      proxy_pass "test_app_url";
      proxy_redirect     off;

      proxy_set_header   Host             $host;
      proxy_set_header   X-Real-IP        $remote_addr;
      proxy_set_header   X-Forwarded-For  $proxy_add_x_forwarded_for;

      client_max_body_size       10m;
      client_body_buffer_size    128k;

      proxy_connect_timeout      90;
      proxy_send_timeout         90;
      proxy_read_timeout         90;

      proxy_buffer_size          4k;
      proxy_buffers              4 32k;
      proxy_busy_buffers_size    64k;
      proxy_temp_file_write_size 64k;

    }

    keepalive_timeout 5;
    client_max_body_size <%= ENV['NGINX_CLIENT_MAX_BODY_SIZE'] || 1 %>M;
  }
}
```

But when nginx starts I got an error:

```
nginx: [emerg] bind() to 0.0.0.0:80 failed (13: Permission denied)
```

Any ideas what can be a problem here?

**Comment by beanieboi** [comment]
Heroku assigns you a random port. you need to listen on `listen <%= ENV["PORT"] %>;` and not `listen 80;`

**Comment by dawidlenkiewicz** [comment]
Thank you very much~

**Comment by mandarvaze** [comment]
I'm also facing similar problem. 

When I use `listen <%= ENV["PORT"] %>;`, I get "Address already in use" because `gunicorn` is also listening on `$PORT`

Relevant log lines :
```
buildpack=nginx at=start-app cmd=gunicorn --paste etc/production.ini -c config/gunicorn.conf.py --bind :50463 
[INFO] Starting gunicorn 19.9.0
[INFO] Listening at: http://0.0.0.0:50463 (18)
buildpack=nginx at=app-initialized 
buildpack=nginx at=nginx-start 
nginx: [emerg] bind() to 0.0.0.0:50463 failed (98: Address already in use) 
```

@beanieboi Any suggestions ?

**Comment by beanieboi** [comment]
@mandarvaze It really depends on your application and what you want to achieve. Do you want to use nginx as a reverse proxy? if so you need to configure gunicorn to listen on a socket instead and tell nginx to proxy to that socket.

**Comment by mandarvaze** [comment]
@beanieboi Thanks.

Yes, that is the requirement. But I could not find any documentation on how to do that.
I'll search for "heroku nginx reverse proxy gunicorn".

If you have a specific resource that can be useful, please share.

**Comment by beanieboi** [comment]
this setup is not specific to heroku. you might want to search for "nginx gunicorn reverse proxy". unfortunately i don't use python myself and can't give you more help on this

**Comment by mandarvaze** [comment]
I was able to make it work locally (macOS)
I've added `127.0.0.1 api.mydomain.com` to my local `/etc/hosts`
So when I enter api.mydomain.com in the local browser, I indeed see `myapp.herokuapp.com/api` in the browser. (It works)

But with heroku it isn't working.

Right now, I'm seeing `No such app (There's nothing here, yet.)` as mentioned [here](https://stackoverflow.com/q/51432288)

So via `CNAME` record, I'm reaching heroku (Cause the above message is heroku branded)
But heroku is unable to "forward" it correctly.

I'll keep looking.
